client/src/main/java/io/pravega/client/byteStream/ByteStreamReader.java

```diff
@@ -0,0 +1,119 @@
+/**
+ * Copyright (c) 2018 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.client.byteStream;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.nio.channels.AsynchronousChannel;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Allows for reading raw bytes from a segment. This class is designed such that it can be used with
+ * or without blocking. To avoid blocking use the {@link #onDataAvailable()} method to make sure to
+ * only call {@link #read(byte[])} when there is data {@link #available()}.
+ *
+ * It is safe to invoke methods on this class from multiple threads, but doing so will not increase
+ * performance.
+ */
+public abstract class ByteStreamReader extends InputStream implements AsynchronousChannel, AutoCloseable {
+
+    /**
+     * Returns the current byte offset in the segment. This call does not block.
+     */
+    public abstract long getOffset();
+
+    /**
+     * Jumps to the provided offset. Future read calls will read from this position. This makes a
+     * synchronous RPC to the server to validate the offset provided.
+     * 
+     * @param offset The offset to jump to.
+     * @throws InvalidOffsetException If the offset provided does not exist in the segment.
+     */
+    public abstract void jumpToOffset(long offset) throws InvalidOffsetException;
+
+    /**
+     * Returns the number of bytes that can be read without blocking. If the number returned is > 0
+     * then a call to {@link #read(byte[]))} will return data from memory without blocking. If the
+     * number returned is 0 then {@link #read(byte[]))} will block If -1 is returned this indicates
+     * the end of the stream has been reached and a call to {@link #read(byte[])} will return -1.
+     * 
+     * @see java.io.InputStream#available()
+     */
+    @Override
+    public abstract int available();
+
+    /**
+     * This make a RPC to the server to fetch the highest offset in the segment.
+     */
+    public abstract long fetchTailOffset();
```

Comment:
I am a little wary of using the word 'length' in our external APIs in a way that is inclusive of data that has been truncated from the beginning of a segment. @fpj What do you think?

Reply:
Agreed, length is not accurate in the presence of truncation, but at the same time, a user could think that an offset of 1000 means that there are 1001 bytes in the stream. I have a slight preference for `offset` because we can define it as a number representing a position in the stream with no relation to the current length. We may want to clarify it in the javadoc in any case.

Reply:
Added some text.

Reply:
Good with me. One really minor point is that it should be `an RPC` .